src/EditorFeatures/Core/EditorConfigSettings/Updater/SettingsUpdateHelper.cs

```diff
@@ -214,87 +260,126 @@ internal static partial class SettingsUpdateHelper
-
-                // Replacing characters in the header with the regex equivalent.
-                fileName = fileName.Replace(".", @"\.");
-                fileName = fileName.Replace("*", ".*");
-                fileName = fileName.Replace("/", @"\/");
-
-                return (fileName, splicedFileExtensions);
-            }
+                var curLineText = curLine.ToString();
+                if (IsOption(curLineText))
+                {
+                    var (untrimmedKey, key, value, severity, comment) = GetOptionParts(curLineText);
 
-            static bool IsNotEmptyOrComment(string currentLineText)
-            {
-                return !string.IsNullOrWhiteSpace(currentLineText) && !currentLineText.Trim().StartsWith("#", StringComparison.OrdinalIgnoreCase);
-            }
+                    // Verify the most recent header is a valid header
+                    if (IsValidHeader(mostRecentHeader, lastValidHeader) &&
+                        string.Equals(key, optionName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // We found the rule in the file -- replace it with updated option value.
+                        textChange = new TextChange(curLine.Span, $"{untrimmedKey}={optionValue}{comment}");
+                    }
+                }
+                else if (IsHeader(curLineText))
+                {
+                    mostRecentHeader = curLine;
+                    if (ShouldSetAsLastValidHeader(curLineText, out var mostRecentHeaderText))
+                    {
+                        lastValidHeader = mostRecentHeader;
+                    }
+                    else
+                    {
+                        var (fileName, splicedFileExtensions) = GetHeaderParts(mostRecentHeaderText);
+                        if ((relativePath.IsEmpty() || new Regex(fileName).IsMatch(relativePath)) &&
+                            HeaderMatchesLanguageRequirements(language, splicedFileExtensions))
+                        {
+                            lastValidHeader = mostRecentHeader;
+                        }
+                    }
+                }
 
-            static bool HeaderMatchesLanguageRequirements(Language language, string[] splicedFileExtensions)
-            {
-                return IsCSharpOnly(language, splicedFileExtensions) || IsVisualBasicOnly(language, splicedFileExtensions) || IsBothVisualBasicAndCSharp(language, splicedFileExtensions);
+                // We want to keep track of how far this (valid) section spans.
+                if (IsValidHeader(mostRecentHeader, lastValidHeader) && IsNotEmptyOrComment(curLineText))
+                {
+                    lastValidHeaderSpanEnd = curLine;
+                    if (lastValidSpecificHeader != null && mostRecentHeader.Equals(lastValidSpecificHeader))
+                    {
+                        lastValidSpecificHeaderSpanEnd = curLine;
+                    }
+                }
             }
```

Comment:
i admit my eyes glazed over with all the string processing.  i really don't have any idea if this is right or not.  Note: we have a generalized pattern matching system in Roslyn for matching over sequences of elements.  it might be worthwhile seeing if we could use that.  but it might also be overkill.  

my primary concern here is that there seems to be a bunch of complexity here that would likely be good to extrac tout into small types that are responsibel for just one thing.  i.e. this is the type that just finds the relevant portions of a chunk of hte editorconfig file.

Reply:
yep, I hope to do a big refactor as part of https://github.com/dotnet/roslyn/issues/52735 and clean all this up